Implement RGate in Rust

jlapeyre opened the pull request:
This completes all tasks of implementing `RGate` in Rust rather than Python.

This replaces 
* #12507

#12507 has been modified and split into the present PR and #12651

jlapeyre committed:
Implement RGate in Rust

ElePT commented on test/python/circuit/test_rust_equivalence.py:
```diff
@@ -21,7 +21,7 @@
 from qiskit.circuit.library.standard_gates import get_standard_gate_name_mapping
 
-SKIP_LIST = {"rx", "ry", "ecr"}
+SKIP_LIST = {"rx", "ry", "ecr", "r"}
```
Why add the `r` gate to the SKIP_LIST? All others haven't been implemented yet, but `r` has an implementation.

jlapeyre replied:
Good question. I added and removed several times while debugging. I'll recheck.

jlapeyre replied:
Fixed in [`7eef0f3` (#12662)](https://github.com/Qiskit/qiskit/pull/12662/commits/7eef0f3d8292f12ebbdcb04b1fb3834e5141ea1d). It failed because the decomposition had an error.

ElePT commented on crates/circuit/src/operations.rs:
```diff
@@ -954,7 +959,41 @@ impl Operation for StandardGate {
+                    )
+                    .expect("Unexpected Qiskit python bug"),
+                )
+            }),
```
I think this could be more readable if we used helper functions like [`multiply_param`](https://github.com/Qiskit/qiskit/blob/6974b4500f6c716b407b599e5cec80afbb757516/crates/circuit/src/operations.rs#L983), and maybe add one for adding a scalar to a parameter in the same fashion. But I am fine merging the implementation as-is and beautifying it later.

jlapeyre replied:
My first iteration did not fit with the helper function, nor one for `add`. But it's changed a lot since I last tried. I'll revisit it; readability and encapsulation is important.

jlapeyre replied:
I did something similar to your suggest in [`49c8dcd` (#12662)](https://github.com/Qiskit/qiskit/pull/12662/commits/49c8dcd69f8c31dd0dd48be4425e9af6b3357dd4). This factors out cloning a `Param`. This is only used once now, but can be used whenever a `Param` needs to be cloned. I find the code much more readable at the call point.

Regarding factoring out `add_param`: There are pros and cons to doing this. Disadvantages: The branch with `-*phi + PI / 2.0` is more readable if it is not nested calls. In fact, this made it slightly easier for me to spot the error `-*phi + 1.0`.

Another one is that it requires two match statements rather than one, so `-*phi + PI / 2.0`  won't be optimized. I'm not sure, but I doubt this will matter in the future for performance.

But, I'm on the fence. Currently, the second match arm would be much more readable with your suggestion. I can implement and use `add_param` quickly if you prefer. I'm inclined to do it just to look at it.

jlapeyre replied:
Implemented your suggestion in [`07fec37` (#12662)](https://github.com/Qiskit/qiskit/pull/12662/commits/07fec379319090b0b065b2589b63ddc75b7430c4). I'm not 100% sure there isn't a non-negligible efficiency hit. But I think it's easier to read and will be easier to build on.

ElePT replied:
I see your point, I think we could merge the code as-is and once all Rust standard gates are in place, maybe run the benchmarks with/without `add_param` to see if there is indeed a significant performance hit.

ElePT reviewed (commented):
Thanks for the PR! it looks pretty good, I just left a couple of comments.

ElePT commented on crates/circuit/src/operations.rs:
```diff
@@ -954,7 +959,41 @@ impl Operation for StandardGate {
+                    }
+                    Param::Obj(_) => unreachable!(),
+                };
+                let (phi_expr1, phi_expr2) = match &params[0] {
```
Isn't phi the second parameter:

```suggestion
                let (phi_expr1, phi_expr2) = match &params[1] {
```

jlapeyre committed:
Update crates/circuit/src/operations.rs

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

jlapeyre committed:
Fix error in decomposition of RGate

There is an error in the expression for decomposition of the R gate
in the port to Rust.

This fixes the error and re-enables the skipped test that failed because
of the incorrect expression.

jlapeyre committed:
Factor cloning the Param enum in Rust

To clone the enum, each variant must be handled separately.  This is currently used once,
but can be used each time a `Param` is cloned. In case more work needs to be done within
match arms, one might choose not to use this function, but rather clone in each of these
arms.

jlapeyre committed:
Run cargo fmt

jlapeyre committed:
Implement and use addition for enum Param

This handles `Float` and `ParameterExpression` variants uniformly.

jlapeyre committed:
Merge branch 'main' into implement-r-gate

ElePT reviewed (approved):
Thanks a lot @jlapeyre! I see that there might be a trade-off between readability and some performance here, but given that the RGate is used in #12650, I would merge it now to unblock downstream PRs and leave any further optimizations for a follow-up, as I mentioned in my other comment.
